### PR TITLE
(PUP-9638) Verify CA fingerprint 

### PR DESCRIPTION
**Pull request by joshcooper**
If `ca_fingerprint` puppet setting is specified, then the agent will verify the downloaded CA certificate and reject it if its fingerprint does not match.

### CODE REVIEW

**Comment by joshcooper** [comment]
digest algorithm needs to be configurable as well. The existing `digest_algorithm` defaults to `md5` though, while the current fingerprint code uses SHA256. Also `agent` allows digest to be specified as a CLI option, while `bootstrap` does not.

**Comment by joshcooper** [comment]
Blocked on https://github.com/puppetlabs/puppet/pull/7579

File: spec/unit/ssl/state_machine_spec.rb [hunk 1]
```diff
@@ -318,6 +344,32 @@ def expect_lockfile_to_contain(pid)
...
+    context 'when verifying CA cert bundle' do
+      before :each do
+        allow(cert_provider).to receive(:load_cacerts).and_return(nil)
+        stub_request(:get, %r{puppet-ca/v1/certificate/ca}).to_return(status: 200, body: cacert_pem)
+        allow(cert_provider).to receive(:save_cacerts)
+      end
+
+      it 'verifies CA cert bundle if a ca_fingerprint is given case-insensitively' do
+        Puppet[:log_level] = :info
+        machine = described_class.new(digest: 'SHA256', ca_fingerprint: 'caacf69bbbcdad9dbcda92dd2da3608b639d1aea4c314d6cc6823cdb32d8e0f8')
```
Comment (jtappa) [inline spec/unit/ssl/state_machine_spec.rb]: Are they always only ever going to be SHA256? Is it assumed that if a SHA256 works, so would a SHA512, or any other algorithm given, or should those also be being tested here?

Comment (joshcooper) [reply spec/unit/ssl/state_machine_spec.rb]: It'd be good to verify if a non-default digest is passed, that it passes as well. I'll add a test for that.

Note it's not currently possible to pass the digest from the command line,

File: spec/unit/ssl/state_machine_spec.rb [hunk 2]
```diff
@@ -496,6 +548,8 @@ def expect_lockfile_to_contain(pid)
         allow(cert_provider).to receive(:load_private_key).and_return(private_key)
         allow(cert_provider).to receive(:load_client_cert).and_return(cert_fixture('tampered-cert.pem'))
 
+        ssl_context = Puppet::SSL::SSLContext.new(cacerts: [cacert], crls: [crl])
+        state = Puppet::SSL::StateMachine::NeedKey.new(machine, ssl_context)
```
Comment (kris-bosland) [inline spec/unit/ssl/state_machine_spec.rb]: How did this test pass before this change?

Comment (joshcooper) [reply spec/unit/ssl/state_machine_spec.rb]: This constructs the `SSLContext` and `NeedKey` state the same way that it did prior, with only the root CA and CRL.

File: spec/unit/ssl/state_machine_spec.rb (hunk 1 above)
Comment (joshcooper) [reply spec/unit/ssl/state_machine_spec.rb]: Added test in https://github.com/puppetlabs/puppet/compare/44241e0b41ef6c727ad462366f20e95e742f0acb..43472eef86e440ee3119f8d5c041bf2fde3e329e

**Pull request merged**